examples/rvc-app/rvc-common/src/rvc-device.cpp

```diff
@@ -104,7 +106,39 @@ void RvcDevice::HandleOpStatePauseCallback(Clusters::OperationalState::GenericOp
 
 void RvcDevice::HandleOpStateResumeCallback(Clusters::OperationalState::GenericOperationalError & err)
 {
-    auto error = mOperationalStateInstance.SetOperationalState(to_underlying(OperationalState::OperationalStateEnum::kRunning));
+    // This method is only called if the device is in a resume-compatible state, i.e. `Charging`, `Docked` or `Paused`.
+    auto error = CHIP_NO_ERROR;
+    switch (mOperationalStateInstance.GetCurrentOperationalState())
+    {
+    case to_underlying(RvcOperationalState::OperationalStateEnum::kCharging):
+    case to_underlying(RvcOperationalState::OperationalStateEnum::kDocked): {
+        if (mRunModeInstance.GetCurrentMode() == RvcRunMode::ModeCleaning ||
+            mRunModeInstance.GetCurrentMode() == RvcRunMode::ModeMapping)
+        {
+            error = mOperationalStateInstance.SetOperationalState(to_underlying(OperationalState::OperationalStateEnum::kRunning));
+        }
+        else
+        {
+            err.Set(to_underlying(OperationalState::ErrorStateEnum::kCommandInvalidInState));
+            return;
+        }
+    }
+    break;
+    case to_underlying(OperationalState::OperationalStateEnum::kPaused): {
+        if (mRusumeState == to_underlying(RvcOperationalState::OperationalStateEnum::kSeekingCharger))
+        {
+            error = mOperationalStateInstance.SetOperationalState(
+                to_underlying(RvcOperationalState::OperationalStateEnum::kSeekingCharger));
+        }
+        else
+        {
+            error = mOperationalStateInstance.SetOperationalState(to_underlying(OperationalState::OperationalStateEnum::kRunning));
+        }
+    }
+    break;
+    }
+
+    // populate the response.
```

Comment:
I am not sure this comment helps me read the code better.

I wonder if we can have a template function so we can call something like:

```cpp
SetError(
    err, 
    (error == CHIP_NO_ERROR) 
        ? OperationalState::ErrorStateEnum::kNoError
        : OperationalState::ErrorStateEnum::kUnableToCompleteOperation
);
```

We have a lot of to_underlying around here

Reply:
This was briefly mentioned at some point when we were working an all of these aliased/derived clusters. I don't think that this is for this example to solve. We should have a discussion on how we want to fix this for all derived cluster and implement a holistic solution.